packages/web3-core/types/index.d.ts

```diff
@@ -410,9 +413,14 @@ export interface LogsOptions {
 
 export type BlockNumber = string | number | BN | BigNumber | 'latest' | 'pending' | 'earliest' | 'genesis';
 
+export interface AbstractProvider extends EventEmitter {
```

Comment:
I would like to consider the following points
- Using the `extends EventEmitter`.
- Remove `extends EventEmitter` (Leaving only the `send` function in the interface).
- Directly define the `on`, `once`, and `removeListener` functions as well as the `WebsocketProviderBase`.

Reply:
My `$0.2` is to separate the core functionality from the subscription functionality. Not all custom providers support subscription features by default. Classes can implement multiple interfaces in TypeScript, so it shouldn't be problematic to export two interfaces.

I am not an active contributor to this code-base. This is my opinion only.

Reply:
Hey @odanado thanks for this PR :) It LGTM. For your 3 points, perhaps just on/once/removeListener is sufficient as the "bare minimum".

What are the types of `WebsocketProviderBase`?

Also, in light of [recent updates to EIP-1193](https://github.com/ethereum/EIPs/issues/2319#issuecomment-610610675), there is a new `request` method that would be great to add. I don't think having `sendAsync` would hurt either even though it's deprecated it has more long term support.

Reply:
@ryanio  Thank you for the comment.
I didn't know the `request` method of EIP-1193.

I thought that the interface of the `AbstractProvider` should be an only of `send` and `sendAsync`.
The reasons are as follows.
- The status of EIP-1193 is a draft.
- EIP-1193 is not yet fully implemented in Metamask

I think it should define a provider that has methods such as on/once/removeListener/request separately from `AbstractProvider`.